### ISSUE TITLE
Inexplicable artefacts in isochrones

### ISSUE THREAD

**Issue by mesozoic-drones** [valhalla/valhalla#3212]
Hello! I've noticed that in multiple cases valhalla generates rather strange isochrones with artefacts. Is this a bug or do I need to tune API parameters in some way?

Here are some example screenshots. 

The most eye-catching case. Artefacts on water:
<img width="500" alt="Screenshot 2021-07-12 at 15 47 58" src="https://user-images.githubusercontent.com/54934129/125652219-955be151-6f0a-4082-bb29-07503dfbdd0b.png">

Another example of artefacts on water:
<img width="500" alt="Screenshot 2021-07-12 at 15 48 24" src="https://user-images.githubusercontent.com/54934129/125652553-69c99f04-b6e5-4051-8dea-5328472abc32.png">

Artefacts on park:
<img width="500" alt="Screenshot 2021-07-12 at 19 43 10" src="https://user-images.githubusercontent.com/54934129/125652657-bc616ee7-66ef-4fb2-a8a0-88667a78e952.png">

Artefact near building:
<img width="500" alt="Screenshot 2021-07-12 at 15 47 13" src="https://user-images.githubusercontent.com/54934129/125652906-42ff228c-8964-4e4c-a841-a821989281d3.png">

Additional information:
- Valhalla uses tiles with the same OSM planet version as the vector tiles on map. 
- I double-checked that there are no OSM ways (like piers, paths etc) that are not shown on map but present in valhalla tiles.
- "generalize" is set to 2.5 meters. **If I use the default, artefacts become more simplified but are still present.**
- All isochrones are built in pedestrian mode.

Thank you in advance!

**Comment by kevinkreiser** [comment]
@mesozoic-drones the way it works underneath is that we make what is essentially a raster image and then we generate vector contours from that rasterization. that step of taking something discretized and adding precision to it often creates issues like the one in your last image where there is a cusp near where some geometry is, basically the center of the pixel that got marked by that little piece of road is not right on the road and so when it gets turned into vector it offsets the vector contour.

the other artifacts are basically that same process but occur where you have holes or where the gradient between pixels values is really high. when the gradient is high we need to interpolate vector lines in a tight space and the way numerical precision works out we get those values that we get :smile: again though the reason the vectors that come out arent smooth is because the pixels that are set in the raster have a jagged edge. basically the underlying data looks like this:

![image](https://user-images.githubusercontent.com/697548/125675850-9969527f-4a8a-4975-a630-c0c7e824bbec.png)

so here where there is an edge between black and white the vector will not be smooth it will be wavey. The only way to fix this is to increase the resolution of the raster, which we totally can do but it makes the algorithm for creating the vectors and also creating the raster slower. If you want to play with it i can show you wehre to change the code to test it out.

**Comment by mesozoic-drones** [comment]
Thank you for explanation! Yes, it would be nice if you show the place where to look.

**Comment by nilsnolde** [comment]
would be nice to have some experiments on this, I agree it can be weird shapes esp for fine-grained isochrones such as pedestrian where accuracy should matter more.. 

if you look here:
https://github.com/valhalla/valhalla/blob/nn-tree-view/src/thor/isochrone.cc#L140-L148

`grid_size` is what determines the size of each cell, i.e. resolution. right now you can see it'll be fixed to 0.001 - 0.005, where the 0.001 could be questionable: it drives the resolution for small isochrones such as most pedestrian ones (even a TODO there to review this again :D). it definitely seems to me there should be more care taken when deciding on (at least) on the lower grid size bound as the major driver for those artifacts.

I haven't really thought about it either how make a better decision for the grid size, but maybe you develop a few looking at the code (also keep in mind we don't _have to_ necessarily hard-code all decision metrics, could be tuneable via global config or api param). very happy to discuss this, bothers me a little too, though apparently not to the critical point;).

**Comment by mesozoic-drones** [comment]
Thank you! I'll research this problem and try to develop suitable solution as soon as I get some time off.

**Comment by mickgrig** [comment]
@kevinkreiser 
> @mesozoic-drones the way it works underneath is that we make what is essentially a raster image and then we generate vector contours from that rasterization. that step of taking something discretized and adding precision to it often creates issues like the one in your last image where there is a cusp near where some geometry is, basically the center of the pixel that got marked by that little piece of road is not right on the road and so when it gets turned into vector it offsets the vector contour.
> 
> the other artifacts are basically that same process but occur where you have holes or where the gradient between pixels values is really high. when the gradient is high we need to interpolate vector lines in a tight space and the way numerical precision works out we get those values that we get smile again though the reason the vectors that come out arent smooth is because the pixels that are set in the raster have a jagged edge. basically the underlying data looks like this:
> 
> ![image](https://user-images.githubusercontent.com/697548/125676070-f5e21b05-e426-442b-b1c3-5e11957525e7.png)
> 
> so here where there is an edge between black and white the vector will not be smooth it will be wavey. The only way to fix this is to increase the resolution of the raster, which we totally can do but it makes the algorithm for creating the vectors and also creating the raster slower. If you want to play with it i can show you wehre to change the code to test it out.

Hi, Kevin!

Let me clarify the details of isochrones' implementation a little bit more. In your post you mentioned some kind of raster image. As far as I could understand, you meant isotile, i.e. 2-dimensional grid of subtiles. So, first of all you build this isotile, then apply Dijkstra algorithm, and while it works you fill up cells of this isotile with some rounded values. After that all the contours are built via this filled isotile.

I mostly wonder is it really necessary to use such kind of grid for this task? I mean, can we just directly use edges and vertices after Dijkstra procedure to calculate all the points for isocounturs? And after that just connect them to build countur. Did you consider similar approaches? Maybe it could allow to get rid of these mentioned artefacts.

I would appreciate any information and feedback you could give! Thanks!

**Comment by nilsnolde** [comment]
> And after that just connect them to build countur

that "just" is the killer I think and may be why valhalla went with the gridded approach. have a look at alpha shapes/concave hulls, nasty business;) for the current approach it's the cell resolution, for alpha shapes it's the input for the delaunay (or similar), basically how concave do you want to get? a pretty non-trivial question and I think if we manage to derive proper resolution sizes the current method would produce so much more accurate geometries compared to classic alpha shapes. but that's just my 2c based on a bit of experience and gut feeling;)

**Comment by kevinkreiser** [comment]
while i do think the idea of not using a raster to generate the contours is an interesting one (i love the types of algorithms that would be involved), i do think @nilsnolde is right. what you would have to do is:

1. still pick some resolution along the shapes of the edges at which to drop "points" into a point cloud of `lat,lon,metric(seconds)`
2. compute the delanay triangulation of this point cloud: i have previously implemented such a thing and it is definitely not trivial. add to that the problem that its run time increases with the density of the network. the raster-based contouring algorithm doesnt get costlier based on the density of the network
3. run the contouring algorithm over the tesselation: its probably possible to write a sweep algorithm that for each value of the metric (each user specified contour time value) we keep track of the intersections along the edges of adjacent triangles and connect them as we go... or we could just loop over all edges of all triangles and find the intersection points which match the input values for time and then connect them in a subsequent step
4. connect intersections to make rings: take all of those interesctions and connect them in a ring, if we didnt figure out a way to make the sweep algorithm that keeps track of the ring as it sweeps

at the end of the day i think its possible but i think it will be slower. @nilsnolde and I agree that it'd probably be quicker for us to focus on picking a more appropriate resolution for a given patch of the planet. actually maybe there is even a hybrid approach where we make the point cloud above and then we rasterize after we have it, perhaps doing the rasterization after the expansion of the edges would allow us to make a better judgement about resolution somehow

**Comment by mickgrig** [comment]
Yeah, I see. I can easily imagine that everthing is not so simple as it seems:)  I'm just curious how that kind of problem is supposed to be solved commonly.
By the way, we did some experiments with parameters tuning for current gridded implementation. And apparently it works fast enough with grid size about 0.0002-0.0005 (less than 0.001). But to get more presentable contours you should increase generalization parameter. Currently, you provide possibility to modify it directly or use some kind of "optimal" value, which defines as tilesize * kMetersPerDegreeLat * 0.25. So, for my experiments contours look more presentable with factor 1 (not 0.25). I mean, maybe modifying this multiplication factor is more reasonable than modifying generalize parameter in direct way.
Thanks for your answers, guys!

**Comment by nilsnolde** [comment]
if you could turn those tests into a few example pics, that'd be great:) we (or I) definitely would like small contours to be more accurate (I think the bigger ones are mostly fine). it's easiest to convince with some before/after pics, maybe even some timings to see we're not going down a rabbit hole (even if it's just HTTP API timings, gives a good clue already).

**Comment by mickgrig** [comment]
Yes, sure. Let me double check it and I will post examples here.

**Comment by mickgrig** [comment]
@nilsnolde 
Original isocontours for pedestrians (3min and 30 min). Grid_size = 0.001, gen_factor = 0.25, perf = 20ms.
![image](https://user-images.githubusercontent.com/36229852/130068004-7f6969f9-908e-485d-ab80-1dffc6b5da72.png)
Grid_size = 0.0005, gen_factor = 0.25, perf = 35ms.
![image](https://user-images.githubusercontent.com/36229852/130068267-db487aaf-ab2b-4c9a-a518-22b421fe6898.png)
Grid_size = 0.0002, gen_factor = 0.25, perf = 70ms.
![image](https://user-images.githubusercontent.com/36229852/130068580-ff906626-b397-4700-a438-8f928639bb85.png)
Grid_size = 0.001, gen_factor = 0.5, perf = 20ms.
![image](https://user-images.githubusercontent.com/36229852/130068699-3f4bf9bd-87fa-4f1f-8500-e646e3e12b63.png)
Grid_size = 0.001, gen_factor = 1, perf = 20ms.
![image](https://user-images.githubusercontent.com/36229852/130068753-258e96b7-2c65-42a6-88cd-cdd6b9b240ae.png)
Grid_size = 0.0002, gen_factor = 1, perf = 80ms.
![image](https://user-images.githubusercontent.com/36229852/130068856-16a777aa-ac71-4a19-9387-afdb80e37c43.png)
Grid_size = 0.0005, gen_factor = 1, perf = 30ms.
![image](https://user-images.githubusercontent.com/36229852/130069020-703a76ed-70c3-487c-81ee-01ca75d0d520.png)

I was measuring performance values locally with ab-utility (Apache HTTP server benchmarking tool). The main point here which I could figure out, that the way we can smooth mentioned artefacts is to use fine-grained enough grid with big enough relative generalization. It looks acceptable for pedestrians, but I'm not sure actually how it will be workings for cars, for example.

**Comment by nilsnolde** [comment]
great, thanks @mickgrig , IMO that looks pretty good and if we're able to determine a good decision model for both the generalization and the grid resolution depending on the travel mode and e.g. max_minutes/km. of course that's a problem if you do, say, [1, 10, 100] mins, we can't scale the grid according to 1 min. which is fair, people can always put multiple successive requests instead, it does need to be mentioned in the docs though (even now, though the tight range of resolutions is not playing a huge role currently, hence this issue:))

**Comment by nilsnolde** [comment]
fwiw, for the shown examples the combo `Grid_size = 0.0002, gen_factor = 1, perf = 80ms` looks pretty nice:) this whole thing will always be a trade-off I guess between resolution/accuracy, performance and (minor) output size.

when we get to an implementation we'd have to somewhat aim to optimize for certain settings I think, so it'd be interesting what the p50/p90 are for max isochrone distances/minutes per request. gut feeling it's 50/100 mins for car and maybe 10/30 for pedestrian, so we shouldn't drop in performance too badly for those settings. I do think that it'd be entirely fine to be half as performant for smaller isochrones like 10 mins for car, 5 mins for pedestrian with the upside of much more accurate geometries.

any comments on those thoughts @kevinkreiser ?

**Comment by kevinkreiser** [comment]
i agree that we dont care about performance of small isochrones. p50 is probably a 30min car isochrone if this is slower by 10ms no one will care for the increase in quality. the p90 (probably like 1+ hour car isochrone) should hopefully not see much of a performance decrease. even though these request types are not common we must avoid an expoential performance per complexity curve, otherwise these p90 requests turn into self-made ddos attacks
